godot-macros/src/util.rs

```diff
@@ -220,6 +226,34 @@ pub(crate) fn validate_impl(
     }
 
     // impl Trait for Self -- validate Self
+    validate_self(original_impl, attr)
+}
+
+/// Validates that the declaration is the of the form `impl Trait for
+/// SomeType`, where the name of `Trait` ends in `Virtual`.
+pub(crate) fn validate_trait_impl_virtual(
+    original_impl: &Impl,
+    attr: &str,
+) -> ParseResult<(Ident, Ident)> {
+    let trait_name = original_impl.trait_ty.as_ref().unwrap(); // unwrap: already checked outside
+
+    // Validate trait
+    if !extract_typename(trait_name).map_or(false, |seg| seg.ident.to_string().ends_with("Virtual"))
+    {
+        return bail(
+            format!("#[{attr}] for trait impls requires a virtual method trait (trait name should end in 'Virtual')"),
+            original_impl,
+        );
+    }
+
+    // Validate self
+    validate_self(original_impl, attr).map(|class_name| {
+        let trait_name = extract_typename(trait_name).unwrap(); // unwrap: already checked in 'Validate trait'
```

Comment:
`extract_typename(trait_name).unwrap()` should be declared at the beginning of the function, this is already the 2nd time you use it. 

This would also simplify the `map_or` above.

Reply:
:+1: